### PR TITLE
ecomode_addon-2

### PR DESCRIPTION
**Pull request by masterwishx**
add-on for #2637,#2680 ECO/Bypass for Eaton Models

@jimklimov In this new pr we can:

You think we can add ESS stuff also, however
Only enterprise UPSes have it?

- [ Done ] We can add also post limits not passed in debug when check range eco/bypass func.

- [ Done ] added variables: 
```
input.bypass.switchable
input.transfer.bypass.overload
input.transfer.bypass.outlimits
```

- [ Done ] ESS added to usbhid-ups.c

- [ Done ] Refactoring : `eaton_input_bypass_check_range()` , `eaton_input_eco_check_range()`

CC @jimklimov please confirm if it needed .

Also for now can't leave eco mode by commands but maybe Becouse my ussue with timeout, not sure...

### CODE REVIEW

**Comment by masterwishx** [comment]
@arnaudquette-eaton Very sorry to bother you again ,maybe you can find info about:  

`UPS.PowerConverter.Input.[2].Switchable`

**Comment by arnaudquette-eaton** [comment]
no need to be sorry @masterwishx , I'm always happy to help ;)

Desc: Item of Bypass AC Input collection
Transfer on automatic Bypass enabled

RW

0: The automatic transfer on bypass is not allowed
1: The automatic transfer on bypass is allowed. (Refer to UPS.PowerConverter.Input.[2].OutOfToleranceTransferEnable and UPS.PowerConverter.Input.[2].OverloadTransferEnable for additional bypass rules)

Hope it helps

**Comment by masterwishx** [comment]
> Transfer on automatic Bypass enabled

Thanks a lot , so its like `UPS.PowerConverter.Input.[2].ForcedTransferEnable` but like for UPS itself bypass autoswitching with rules?

as you know we now have ECO/Bypass enabled with your help for info!

the only problem i cant exit from ECO to bypass by r/w values or commands bypass.on/off , but maybe its local issue for my UPS with timeouts ....

Also will try to add function for auto enter to ECO mode ( bypass.on + ECO + bypass.off) but still cant test exit from ECO  :( 
Thanks again for your help

**Comment by masterwishx** [comment]
> Transfer on automatic Bypass enabled

added values if you think they usefull ?

**Comment by masterwishx** [comment]
CC @arnaudquette-eaton can you please check and confirm that we can exit from ECO mode by : 
`"UPS.PowerConverter.Input.[2].SwitchOffControl"=1`  `input.bypass.switch.off`
or maybe some other command ?  (cant exit from ECO by this command, but maybe its my local issue ?!?)

**Comment by masterwishx** [comment]
From https://github.com/networkupstools/nut/pull/2684/commits/8afba96366f4ff9a5d1d190d9e8a0aaf9f3edb5a 
Work for function for auto ECO started , if it needed at all ? CC @jimklimov

File: drivers/mge-hid.c [hunk 1]
```diff
@@ -909,21 +908,26 @@ static const char *eaton_input_eco_mode_auto_on_fun(double value)
 	}
 
 	/* Check if input.eco.switchable is normal and set it to 'ECO' */
-   if (strcmp(eco_switchable_str, "normal")) {
+    if (strcmp(eco_switchable_str, "normal")) {
 	    check_eco_result = eaton_input_eco_mode_check_range(value);
```
Comment (masterwishx) [inline drivers/mge-hid.c:912]: @jimklimov please confirm here we should call function to check range becouse : 
when `setvar("input.eco.switchable", "ECO");` its only set variable but not calling function `eaton_input_eco_mode_check_range` like here, same for bypass?

https://github.com/networkupstools/nut/blob/858f615a32aaef3497442eae0155a773953123dd/drivers/mge-hid.c#L1709-L1710

https://github.com/networkupstools/nut/blob/858f615a32aaef3497442eae0155a773953123dd/drivers/mge-hid.c#L783-L789

**Comment by masterwishx** [comment]
@arnaudquette-eaton Sorry again but can you check and confirm that it was done right ?

`UPS.PowerConverter.Input.[5].PresentStatus.Used`
```
info_lkp_t eco_mode_info[] = {
    { 0, "normal", NULL, NULL },
    { 1, "ecomode", NULL, NULL },
    { 2, "essmode", NULL, NULL },
    { 0, NULL, NULL, NULL }
};
```

or we should use :  

```
info_lkp_t eco_mode_info[] = {
    { 0, "!ecomode", NULL, NULL },
    { 1, "ecomode", NULL, NULL },
    { 2, "essmode", NULL, NULL },
    { 0, NULL, NULL, NULL }
};
```

**Comment by masterwishx** [comment]
@arnaudquette-eaton @jimklimov it seems if we return NULL from function its just not return value but proccess the command or r/w variable ?

https://github.com/networkupstools/nut/blob/f99109dee8575bad2768ae498486ea436fbf739e/drivers/mge-hid.c#L1010

**Comment by jimklimov** [comment]
>  it seems if we return NULL from function its just not return value but proccess the command or r/w variable ?

Can't say quickly and reliably; please check the caller logic in `usbhid-ups.c`.

**Comment by masterwishx** [comment]
> > it seems if we return NULL from function its just not return value but proccess the command or r/w variable ?
> 
> Can't say quickly and reliably; please check the caller logic in `usbhid-ups.c`.

I will try to check  , asked becouse user tested ecomode in #2685 was need not able to enter to eco becouse of return NULL in case of no vars of eco limits ...

https://github.com/networkupstools/nut/blob/c83ebd45d22deaa33647c430f47cc99f9d26f206/drivers/mge-hid.c#L711-L733

**Comment by masterwishx** [comment]
> > it seems if we return NULL from function its just not return value but proccess the command or r/w variable ?
> 
> Can't say quickly and reliably; please check the caller logic in `usbhid-ups.c`.

should be fixed in last commit

**Comment by masterwishx** [comment]
@jimklimov please check when you have time if can be merged

**Comment by jimklimov** [comment]
Seems reasonable code-wise, applied a bit of cleaning and waiting for CI to agree :)
Also a good moment for @arnaudquette-eaton to chime in as an expert on subject matter :-)

**Comment by masterwishx** [comment]
@jimklimov can be merged?
Do you want to change essmode to ECO in this or in next pr?

**Comment by jimklimov** [comment]
I guess it would now be pending the discussion results in #2708, whether to remove `status_set("ESS")` (and perhaps "ECO" too), in favor of changing `upsmon` notifications. That does begin to sound more and more like the reasonable way forward.

The rest seems good though, so can merge it anyway and take care of removal and changes in a next PR.

**Comment by masterwishx** [comment]
> in favor of changing `upsmon` notifications. That does begin to sound more and more like the reasonable way forward.

OK. My main task was to add and enable the missing Bypass/ECO modes for Eaton usb driver.
 The rest of the improvements with deep respect hope for your experience :)

**Pull request merged**